internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/DiagnosticsReporter.scala

```diff
@@ -117,67 +117,64 @@ object DiagnosticsReporter {
+              Option(m.invoke(obj, args: _*))
             // See com.sun.tools.javac.api.ClientCodeWrapper.DiagnosticSourceUnwrapper
             val diagnostic = d.getClass.getField("d").get(d)
             // See com.sun.tools.javac.util.JCDiagnostic#getDiagnosticSource
-            val getDiagnosticSourceMethod =
-              diagnostic.getClass.getDeclaredMethod("getDiagnosticSource")
-            val getPositionMethod = diagnostic.getClass.getDeclaredMethod("getPosition")
-            (Option(getDiagnosticSourceMethod.invoke(diagnostic)),
-             Option(getPositionMethod.invoke(diagnostic))) match {
+            val getDiagnosticSource = diagnostic.getClass.getDeclaredMethod("getDiagnosticSource")
+            val getPosition = diagnostic.getClass.getDeclaredMethod("getPosition")
+            (invoke(diagnostic, getDiagnosticSource), invoke(diagnostic, getPosition)) match {
               case (Some(diagnosticSource), Some(position: java.lang.Long)) =>
                 // See com.sun.tools.javac.util.DiagnosticSource
                 val getLineMethod = diagnosticSource.getClass.getMethod("getLine", Integer.TYPE)
-                Option(getLineMethod.invoke(diagnosticSource, new Integer(position.intValue())))
+                invoke(diagnosticSource, getLineMethod, new Integer(position.intValue()))
                   .map(_.toString)
               case _ => None
             }
           } catch {
-            // TODO - catch ReflectiveOperationException once sbt is migrated to JDK7
-            case _: Throwable => None
+            case _: ReflectiveOperationException => None
           }
 
         def getExpression: String =
-          Option(d.getSource) match {
-            case Some(source: JavaFileObject) =>
+          source match {
+            case None => ""
+            case Some(source) =>
               (Option(source.getCharContent(true)), startPosition, endPosition) match {
                 case (Some(cc), Some(start), Some(end)) =>
                   cc.subSequence(start.toInt, end.toInt).toString
                 case _ => ""
               }
-            case _ => ""
           }
 
         getDiagnosticLine.getOrElse(getExpression)
       }
 
-      def fixSource[T <: JavaFileObject](source: T): Option[String] = {
-        try Option(source).map(_.toUri.normalize).map(new File(_)).map(_.getAbsolutePath)
-        catch {
-          case _: IllegalArgumentException =>
-            // Oracle JDK6 has a super dumb notion of what a URI is.  In fact, it's not even a legimitate URL, but a dump
-            // of the filename in a "I hope this works to toString it" kind of way.  This appears to work in practice
-            // but we may need to re-evaluate.
-            Option(source).map(_.toUri.toString)
+      def sourceUri: Option[String] =
+        source map { obj =>
+          val uri = obj.toUri
+          try new File(uri.normalize).getAbsolutePath
```

Comment:
In our current arsenal, a better bet here is `IO.toFile(uri).getAbsolutePath` instead of the try-catch.

See https://github.com/sbt/io/blob/10dd004d3c0baad7e88016751208def76989316a/io/src/main/scala/sbt/io/IO.scala#L130-L131 and http://eed3si9n.com/encoding-file-path-as-URI-reference

Reply:
I'll push a commit - let me know if I correctly did what you meant.